flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClientActor.java

```diff
@@ -144,11 +230,22 @@ else if (message instanceof Terminated) {
 				String msg = "Lost connection to JobManager " + jobManager.path();
 				logger.info(msg);
 				submitter.tell(decorateMessage(new Status.Failure(new Exception(msg))), getSelf());
+				scheduler.cancel();
+				resetContextAndActor();
 			} else {
 				logger.error("Received 'Terminated' for unknown actor " + target);
 			}
 		}
 
+		// ============= No messgaes received in the job manager timeout duration ========
+		else if (message instanceof ReceiveTimeout){
```

Comment:
According to the JavaDocs of `setReceiveTimeout`, it might happen that you receive a `ReceiveTimeout` message even though you've just received another message. You should guard against this.